src/psyclone/tests/utils.py

```diff
@@ -28,3 +50,141 @@ def count_lines(root, string_name):
         if string_name in line:
             count += 1
     return count
+
+
+def walk(parent, my_type):
+    ''' recurse through a tree of objects and return those that are
+    instances of mytype. Assumes that the children of any given node
+    are held in the list node.children '''
+    local_list = []
+    for child in parent.children:
+        if isinstance(child, my_type):
+            local_list.append(child)
+        local_list += walk(child, my_type)
+    return local_list
+
+
+def find_fortran_file(path, root_name):
+    ''' Returns the full path to a Fortran source file. Searches for
+    files with suffixes defined in FORTRAN_SUFFIXES. Raises IOError
+    if no matching file is found. '''
+    name = os.path.join(path, root_name)
+    for suffix in FORTRAN_SUFFIXES:
+        if os.path.isfile(str(name)+"."+suffix):
+            name += "." + suffix
+            return name
+    raise IOError("Cannot find a Fortran file '{0}' with suffix in {1}".
+                  format(name, FORTRAN_SUFFIXES))
+
+
+def compile_file(filename, f90, f90flags):
+    ''' Compiles the specified Fortran file into an object file (in
+    the current working directory) using the specified Fortran compiler
+    and flags. Raises a CompileError if the compilation fails. '''
+
+    # Build the command to execute
+    if f90flags:
+        arg_list = [f90, f90flags, '-c', filename]
+    else:
+        arg_list = [f90, '-c', filename]
+
+    # Attempt to execute it using subprocess
+    import subprocess
+    try:
+        build = subprocess.Popen(arg_list,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.STDOUT)
+        (output, error) = build.communicate()
+    except OSError as err:
+        print "Failed to run: {0}: ".format(" ".join(arg_list))
+        print "Error was: ", str(err)
+        raise CompileError(str(err))
+
+    # Check the return code
+    stat = build.returncode
+    if stat != 0:
+        print output
+        if error:
+            print "========="
+            print error
+        raise CompileError(output)
+    else:
+        return True
+
+
+def code_compiles(api, psy_ast, tmpdir, f90, f90flags):
```

Comment:
If we can integrate the build tests with existing correctness tests then it might be better to pass in the code as a string here rather than the ast (psy_ast), as it will already have been generated within the test. If we are going for different correctness and compile tests then there is no need to change.

Reply:
Now that we look for USE'd modules using the `psy_ast` we cannot replace it with a string. In fact this means that my optimisation to store the generated AST is now valid again but, as it is an optimisation, I'll leave it as it is for now.